File: src/infrastructure/validators/cnpj.ts
import { IValidationResult, ok, fail, onlyDigits } from './result';

const WEIGHTS_1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
const WEIGHTS_2 = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];

const checkDigit = (digits: string, weights: number[]): number => {
  let total = 0;
  for (let i = 0; i < weights.length; i++) {
    total += Number(digits[i]) * weights[i];
  }
  const remainder = total % 11;
  return remainder < 2 ? 0 : 11 - remainder;
};

export const formatCnpj = (d: string): string =>
  `${d.slice(0, 2)}.${d.slice(2, 5)}.${d.slice(5, 8)}/${d.slice(8, 12)}-${d.slice(12, 14)}`;

/** Valida CNPJ por módulo-11 (pesos oficiais) e retorna o valor mascarado quando válido. */
export function validateCnpj(raw: string | null | undefined): IValidationResult {
  const d = onlyDigits(raw ?? '');
  if (!d) return fail('CNPJ vazio');
  if (d.length !== 14) return fail(`CNPJ deve ter 14 dígitos (encontrado ${d.length})`);
  if (d === d[0].repeat(14)) return fail('CNPJ com todos os dígitos iguais');

  const dv1 = checkDigit(d.slice(0, 12), WEIGHTS_1);
  if (dv1 !== Number(d[12])) return fail('Primeiro dígito verificador do CNPJ inválido');

  const dv2 = checkDigit(d.slice(0, 13), WEIGHTS_2);
  if (dv2 !== Number(d[13])) return fail('Segundo dígito verificador do CNPJ inválido');

  return ok(formatCnpj(d));
}

File: app/validators/cnpj.py
"""Validação e normalização de CNPJ (Cadastro Nacional da Pessoa Jurídica)."""
from __future__ import annotations

import re

from .base import ValidationResult

_NON_DIGITS = re.compile(r"\D+")
_WEIGHTS_1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
_WEIGHTS_2 = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]


def clean(raw: str) -> str:
    return _NON_DIGITS.sub("", raw or "")


def format_cnpj(digits: str) -> str:
    """Aplica a máscara 00.000.000/0000-00 a 14 dígitos."""
    return f"{digits[0:2]}.{digits[2:5]}.{digits[5:8]}/{digits[8:12]}-{digits[12:14]}"


def _check_digit(digits: str, weights: list[int]) -> int:
    total = sum(int(d) * w for d, w in zip(digits, weights))
    remainder = total % 11
    return 0 if remainder < 2 else 11 - remainder


def _check_digits(digits: str) -> bool:
    dv1 = _check_digit(digits[:12], _WEIGHTS_1)
    if dv1 != int(digits[12]):
        return False
    dv2 = _check_digit(digits[:13], _WEIGHTS_2)
    return dv2 == int(digits[13])


def is_valid(raw: str) -> bool:
    digits = clean(raw)
    if len(digits) != 14:
        return False
    if digits == digits[0] * 14:
        return False
    return _check_digits(digits)


def validate(raw: str) -> ValidationResult:
    digits = clean(raw)
    if not digits:
        return ValidationResult(False, None, ["CNPJ vazio"])
    if len(digits) != 14:
        return ValidationResult(False, None, [f"CNPJ deve ter 14 dígitos (encontrado {len(digits)})"])
    if digits == digits[0] * 14:
        return ValidationResult(False, None, ["CNPJ com todos os dígitos iguais"])
    if not _check_digits(digits):
        return ValidationResult(False, None, ["Dígitos verificadores do CNPJ inválidos"])
    return ValidationResult(True, format_cnpj(digits), [])
